scripts/caravan/event-handlers/action.lua

```diff
@@ -127,12 +127,59 @@ end
 
 local prefix = "py_caravan_action_number_selection"
 
+local expression_variables = {k=1000, K=1000, m=1000000, M=1000000, g=1000000000, G=1000000000}
+
+
+--- Parses a mathematical expression using MathExp library with predefined variables.
+--- @param expr string
+--- @param vars { [string]: number }?
+--- @return number?
+local function parse_math_expr(expr, vars)
+    if not expr or expr == "" then return nil end
+
+    local ok, result = pcall(function()
+        return helpers.evaluate_expression(expr, vars or expression_variables)
+    end)
+
+    if ok and type(result) == "number" then
+        -- Floor the result to avoid fractions in item counts or time
+        result = math.floor(result)
+
+        return result
+    else
+        return nil
+    end
+end
+
 local function on_confirmed(event)
     local player = game.get_player(event.player_index)
 
     local tags = event.element.tags
     local textfield = event.element.parent[prefix .. "_text_field"]
     local value = tonumber(textfield.text) or 0
+    local parsed_value = parse_math_expr(textfield.text)
+
+    if value ~= parsed_value and parsed_value then
+        textfield.text = tostring(parsed_value)
+
+        local slider = textfield.parent[prefix .. "_slider"]
+        if slider and slider.valid then
+            -- clamp to slider bounds if present
+
+            -- TODO: Try to get the bounds from tags saved at creation
+            local min_v = 0
+            local max_v = 50000 
+            
+            if min_v and max_v then
+                parsed_value = math.max(min_v, math.min(max_v, parsed_value))
```

Comment:
This doesn't match current behavior where you can manually enter a number greater than the slider length.

Reply:
Additionally, negative numbers can be required for circuit conditions

Reply:
removed clamping . Resolved in commit https://github.com/pyanodon/pyalienlife/pull/500/commits/07a11c3bba4b1b70dc47635c17289643d6e1538e